### PR TITLE
Fixes for is_threadedt, --is-threaded

### PR DESCRIPTION
**Pull request by danpoe**

### CODE REVIEW

**Review by tautschnig — APPROVED** [review]
Wishlist items:
1. Explain in the body of the commit message what is being fix/what was broken before.
2. clang-format isn't happy. Part of that is that there ought to be a "clang-format off/on" around the help message.

File: src/analyses/is_threaded.cpp [hunk 1]
```diff
@@ -33,10 +33,8 @@ class is_threaded_domaint:public ai_domain_baset
     locationt from,
     locationt to)
   {
-    // assert(src.reachable);
-
-    if(!src.reachable)
-      return false;
+    INVARIANT(src.reachable,
```
Comment (thk123) [inline src/analyses/is_threaded.cpp:36]: Probably `PRECONDITION` here expresses the intent clearer.

File: src/analyses/is_threaded.cpp [hunk 2]
```diff
@@ -52,10 +50,8 @@ class is_threaded_domaint:public ai_domain_baset
   transform(locationt from, locationt to, ai_baset &ai, const namespacet &ns)
     final override
   {
-    // assert(reachable);
-
-    if(!reachable)
-      return;
+    INVARIANT(reachable,
```
Comment (thk123) [inline src/analyses/is_threaded.cpp:53]: Ditto about `PRECONDITION`

**Review by thk123 — APPROVED** [review]

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 3]
```diff
@@ -1493,6 +1495,8 @@ void goto_instrument_parse_optionst::help()
     // NOLINTNEXTLINE(whitespace/line_length)
     " --reachable-call-graph       show graph of function calls potentially reachable from main function\n"
     " --class-hierarchy            show class hierarchy\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
+    " --show-threaded              show instructions that may be executed by more than one thread\n"
```
Comment (martin-cs) [inline src/goto-instrument/goto_instrument_parse_options.cpp:1499]: This is just documented and undocumented option, right?

File: src/analyses/is_threaded.cpp (hunk 1 above)
Comment (martin-cs) [reply src/analyses/is_threaded.cpp:36]: @thk123 I'm unsure... my belief is that this is enforced by a property of the AI engine; it's really an invariant of that.  Hmmm... maybe it is a precondition...

**Review by martin-cs — APPROVED** [review]
Approve provided we can get the history of this, w.r.t.

https://github.com/diffblue/cbmc/pull/555
and
https://github.com/diffblue/cbmc/commit/a47cd9421fe1cff3683004739db0aeb46d7c9160

I mean ... I think this commit is (finally) closing the door on this ... but please have a scan through and make sure we're not just caught in an ambiguous commit and revert cycle.

**Comment by tautschnig** [comment]
Assigning to @danpoe as various comments by reviewers need to be responded to/acted upon (despite overall approval).

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 3 above)
Comment (danpoe) [reply src/goto-instrument/goto_instrument_parse_options.cpp:1499]: Yes, it's only documenting a previously undocumented option.

File: src/analyses/is_threaded.cpp (hunk 1 above)
Comment (danpoe) [reply src/analyses/is_threaded.cpp:36]: It's both a precondition of the method and a property of the AI engine, thus I used the more general `INVARIANT`. However, thinking about it now, maybe `PRECONDITION` does indeed suit better, and the fact that the AI engine does not apply `transform()` to bottom should be checked by an invariant in `ait` instead.

**Comment by danpoe** [comment]
@martin-cs #555 added a regression test that shows a case where a47cd94 did not mark some concurrent code as threaded. That should safeguard against changing it back.

**Pull request merged**